### PR TITLE
move tinyc to a separate repo and allow installing external dependencency (eg tinyc) from koch / library code

### PR DESCRIPTION
**Pull request by timotheecour**
this PR addresses parts of https://github.com/nim-lang/RFCs/issues/206

## move out tinyc
* created https://github.com/timotheecour/nim-tinyc-archive which extracts all commit history from Nim/tinyc
* removed tinyc from Nim

EDIT: but I'm still allowing for tinyc to be re-installed (unlike https://github.com/nim-lang/Nim/pull/11529 which was closed because there was no way to install it back)

## allow installing external dependency (eg tinyc etc) to Nim via koch
* added `tools/deps.nim` to make it easy to update external dependency:
`proc cloneDependency*(destDirBase: string, url: string, commit = commitHead, appendRepoName = true)`
which is safe to re-run (for eg with different `commit`)

and using it from koch, eg:
```
./koch installdeps tinyc
```
this will clone https://github.com/timotheecour/nim-tinyc-archive at Nim/dist/nim-tinyc-archive at a fixed revision (the current git HEAD of that repo, corresponding to latest state of tinyc in nim's repo before this PR)

* updated paths referring to tinyc
* made the minimum amount of change to make `nim c -d:tinyc compiler/nim.nim` compile again; whether or not `compiler/tccgen.nim` will eventually be removed is **out of scope for this PR**, at least it currently compiles as of this PR in case someone is curious to revive it (perhaps some code can be reused for `zig` as replacement for tinyc (see https://github.com/nim-lang/Nim/pull/13757 and https://github.com/ziglang/zig/issues/596). Note that tinyc doesn't work on OSX (https://stackoverflow.com/questions/3712902/problems-compiling-tcc-on-os-x#answer-9590412) and lacks documentation (https://github.com/nim-lang/Nim/issues/11495)

## after this PR
* move out linenoise using same approach
* and maybe niminst, zig etc, see https://github.com/nim-lang/RFCs/issues/206 for details
* and definitely anything that requires modifications form upstream (via our own fork)

### CODE REVIEW

**Comment by juancarlospaco** [comment]
https://github.com/nim-lang/Nim/pull/11529#issuecomment-505877648

**Comment by ghost** [comment]
@juancarlospaco this PR doesn't remove tinyc unlike yours, it moves tinyc to an external repository and you still will be able to fetch tinyc with `koch`

File: compiler/tccgen.nim [hunk 1]
```diff
@@ -8,12 +8,20 @@
...
 import
-  os, strutils, options, msgs, tinyc
+  os, strutils, options, msgs, tinyc, lineinfos, sequtils
 
-{.compile: "../tinyc/libtcc.c".}
+const tinyPrefix = "dist/nim-tinyc-archive".unixToNativePath
+const nimRoot = currentSourcePath.parentDir.parentDir
+const tinycRoot = nimRoot / tinyPrefix
+when not dirExists(tinycRoot):
+  static: doAssert false, $(tinycRoot, "requires: ./koch installdeps tinyc")
```
Comment (juancarlospaco) [inline compiler/tccgen.nim:17]: https://nim-lang.github.io/Nim/manual.html#pragmas-fatal-pragma ?

Comment (timotheecour) [reply compiler/tccgen.nim:17]: what's the advantage concretely ?

**Pull request merged**